tests/suites/test_suite_nist_kw.function

```diff
@@ -0,0 +1,177 @@
+
+
+    TEST_ASSERT( mbedtls_nist_kw_setkey( &ctx, MBEDTLS_CIPHER_ID_AES,
+                                     key, 8 * sizeof( key ), iswrap ) == 0 );
+    if ( iswrap == 1)
+    {
+        TEST_ASSERT( mbedtls_nist_kw_wrap( &ctx, mode, plaintext, in_len,
+                                      ciphertext, &output_len ) == res );
+        if( res == 0 )
+        {
+            if( mode == MBEDTLS_KW_MODE_KWP )
+                TEST_ASSERT( output_len <= (size_t)in_len + 16 );
+            else
+                TEST_ASSERT( output_len == (size_t)in_len + 8 );
+        }
+    }
+    else
+    {
+         unwrap_ret = mbedtls_nist_kw_unwrap( &ctx, mode,  ciphertext, in_len,
+                                    plaintext, &output_len );
+
+        if( res == 0 )
+            TEST_ASSERT( unwrap_ret == MBEDTLS_ERR_CIPHER_AUTH_FAILED );
+        else
+            TEST_ASSERT( unwrap_ret == res );
+    }
+
+exit:
+    mbedtls_free( ciphertext );
+    mbedtls_free( plaintext );
+    mbedtls_nist_kw_free( &ctx );
+}
+/* END_CASE */
+
+/* BEGIN_CASE */
+void mbedtls_nist_kw_wrap( int cipher_id, int mode,
+                        char *key_hex, char *msg_hex,
+                        char *result_hex )
+{
+    unsigned char key[32];
+    unsigned char msg[512];
+    unsigned char result[528];
+    unsigned char expected_result[528];
+    mbedtls_nist_kw_context ctx;
+    size_t key_len, msg_len, output_len, result_len;
+
+    mbedtls_nist_kw_init( &ctx );
+
+    memset( key, 0x00, sizeof( key ) );
+    memset( msg, 0x00, sizeof( msg ) );
+    memset( result, 0x00, sizeof( result ) );
+
+    key_len = unhexify( key, key_hex );
+    msg_len = unhexify( msg, msg_hex );
+    result_len = unhexify( expected_result, result_hex );
+
+    TEST_ASSERT( mbedtls_nist_kw_setkey( &ctx, cipher_id, key, key_len * 8, 1 ) == 0 );
+
+    /* Test with input == output */
+    TEST_ASSERT( mbedtls_nist_kw_wrap( &ctx, mode, msg, msg_len,
+                 result, &output_len ) == 0 );
+
+    TEST_ASSERT( output_len == result_len );
+
+    TEST_ASSERT( memcmp( expected_result, result, result_len ) == 0 );
+
+exit:
+    mbedtls_nist_kw_free( &ctx );
+}
+/* END_CASE */
+
+/* BEGIN_CASE */
+void mbedtls_nist_kw_unwrap( int cipher_id, int mode,
+                        char *key_hex, char *msg_hex,
+                        char *result_hex )
+{
+    unsigned char key[32];
+    unsigned char msg[528];
+    unsigned char result[528];
+    unsigned char expected_result[528];
+    mbedtls_nist_kw_context ctx;
+    size_t key_len, msg_len, output_len, result_len;
+    int ret;
+
+    mbedtls_nist_kw_init( &ctx );
+
+    memset( key, 0x00, sizeof( key ) );
+    memset( msg, 0x00, sizeof( msg ) );
+    memset( result, 0x00, sizeof( result ) );
+    memset( expected_result, 0x00, sizeof( expected_result ) );
+
+    key_len = unhexify( key, key_hex );
+    msg_len = unhexify( msg, msg_hex );
+
+    if( strcmp( "FAIL", result_hex ) == 0 )
+    {
+        ret = MBEDTLS_ERR_CIPHER_AUTH_FAILED;
```

Comment:
You can pass MBEDTLS_ERR_CIPHER_AUTH_FAILED in the test vector file as a value for a new int parameter, `ret`. Would that work in this case? Then we wouldn't need to do `strcmp` and the code will be smaller (more readable, less logic), and we could more easily add checks for other error types.

Reply:
Not OK. The definition of the algorithm doesn't output the string `"FAIL"`. It outputs either the status “success” and a string of octets, or the status “fail”. In more advanced type systems this could be represented with an option type. The way the library encodes the result is: if the function returns 0 then the status is “success” and the output string is in the output buffer, and if the function returns a nonzero value then the status is “fail”. Please don't make up your own representation for this data. Use the same encoding in the test data as in the library.

Besides this wouldn't carry over to the upcoming change in the test framework (#930) where binary data is no longer transmitted as a hex string, but as binary data.

Reply:
If you look at the expected output from the NIST test vectors, you will see the string "FAIL". this was taken from there. This is same implementation as in our CCM.
However, if this will not work, once #930 is merged, I will change the expected output.